src/d3d9/d3d9_vertex_declaration.cpp

```diff
@@ -205,6 +206,106 @@ namespace dxvk {
+  // D3DDECLTYPE_FLOATm     D3DDECLUSAGE_TEXCOORD       n               D3DFVF_TEXCOORDSIZEm(n)
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_POSITION       1               N / A
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_NORMAL         1               N / A
+
+  if (element.Usage == D3DDECLUSAGE_POSITION && element.Type == D3DDECLTYPE_FLOAT3 && element.UsageIndex == 0)
+    return D3DFVF_XYZ;
+  else if (element.Usage == D3DDECLUSAGE_POSITIONT && element.Type == D3DDECLTYPE_FLOAT4 && element.UsageIndex == 0)
+    return D3DFVF_XYZRHW;
+  else if (element.Usage == D3DDECLUSAGE_BLENDWEIGHT && element.UsageIndex == 0) {
+    switch (element.Type)
+    {
+    case D3DDECLTYPE_FLOAT1:
+      return D3DFVF_XYZB1;
+    case D3DDECLTYPE_FLOAT2:
+      return D3DFVF_XYZB2;
+    case D3DDECLTYPE_FLOAT3:
+      return D3DFVF_XYZB3;
+    case D3DDECLTYPE_FLOAT4:
+      return D3DFVF_XYZB4;
+    default:;
+    }
+  }
+  else if (element.Usage == D3DDECLUSAGE_BLENDINDICES && element.Type == D3DDECLTYPE_UBYTE4 && element.UsageIndex == 0)
+    return D3DFVF_XYZB1;
+  else if (element.Usage == D3DDECLUSAGE_NORMAL && element.Type == D3DDECLTYPE_FLOAT3 && element.UsageIndex == 0)
+    return D3DFVF_NORMAL;
+  else if (element.Usage == D3DDECLUSAGE_PSIZE && element.Type == D3DDECLTYPE_FLOAT1 && element.UsageIndex == 0)
+    return D3DFVF_PSIZE;
+  else if (element.Usage == D3DDECLUSAGE_COLOR && element.Type == D3DDECLTYPE_D3DCOLOR) {
+    switch (element.UsageIndex)
+    {
+    case 0:
+      return D3DFVF_DIFFUSE;
+    case 1:
+      return D3DFVF_SPECULAR;
+    default:;
+    }
+  }
+  else if (element.Usage == D3DDECLUSAGE_TEXCOORD && element.UsageIndex < 8) {
+      DWORD localFvf = 0;
+      do {
+        // Check if bits of format for current UsageIndex are free in the fvf
+        // It is necessary to skip multiple initializations of the bitfield because
+        // returned value is bitwise or-ed to final fvf DWORD.
+        if (D3DFVF_TEXCOORDSIZE1(element.UsageIndex) & fvf != 0)
+          break;
+  
+        // Update max texture's index in fvf
+        DWORD lastMaxTexCount = (fvf >> 8) & 0x7;
+        DWORD currentTexCount = element.UsageIndex + 1;
+        if (lastMaxTexCount < currentTexCount) {
+          texCountPostUpdate = currentTexCount << 8;
+        }
+  
+        if (element.Type == D3DDECLTYPE_FLOAT1)
+          localFvf = D3DFVF_TEXCOORDSIZE1(element.UsageIndex);
+        else if (element.Type == D3DDECLTYPE_FLOAT2)
+          localFvf = D3DFVF_TEXCOORDSIZE2(element.UsageIndex);
+        else if (element.Type == D3DDECLTYPE_FLOAT3)
+          localFvf = D3DFVF_TEXCOORDSIZE3(element.UsageIndex);
+        else if (element.Type == D3DDECLTYPE_FLOAT4)
+          localFvf = D3DFVF_TEXCOORDSIZE4(element.UsageIndex);
+        else
+          break;
+  
+        return localFvf;
+      } while (1);
+    }
+    Logger::warn("D3D9VertexDecl::MapD3DDeclToFvf: Unsupported set of D3DDECLUSAGE_* / D3DDECLTYPE_* / UsageIndex");
```

Comment:
Put this log into the switches, this will be easy when you put them into the helpers too.

Reply:
My intention was to not multiply these logs when one at the end is necessary.
Do you ant to have this log in helpers too?